src/common/walFilter/versions/recordProcessGPDB7.c

```diff
@@ -0,0 +1,452 @@
+                {
+                    const xl_smgr_truncate *const xlrec = mainData;
+                    return &xlrec->rnode;
+                }
+
+                default:
+                    THROW_FMT(FormatError, "unknown Storage record: %d", info);
+            }
+
+        case RM7_HEAP2_ID:
+            switch (info)
+            {
+                case XLOG_HEAP2_CLEANUP_INFO:
+                    return (const RelFileNode *) mainData;
+
+                case XLOG_HEAP2_NEW_CID:
+                {
+                    const xl_heap_new_cid *const xlrec = mainData;
+                    return &xlrec->target;
+                }
+
+                default:
+                    break;
+            }
+            break;
+
+        case RM7_BTREE_ID:
+            if (info == XLOG_BTREE_REUSE_PAGE)
+                return (const RelFileNode *) mainData;
+            break;
+
+        case RM7_GIN_ID:
+            switch (info)
+            {
+                case XLOG_GIN_SPLIT:
+                case XLOG_GIN_UPDATE_META_PAGE:
+                    return (const RelFileNode *) mainData;
+
+                default:
+                    break;
+            }
+            break;
+
+        case RM7_GIST_ID:
+            if (info == XLOG_GIST_PAGE_REUSE)
+                return (const RelFileNode *) mainData;
+            break;
+
+        case RM7_SEQ_ID:
+            if (info == XLOG_SEQ_LOG)
+                return (const RelFileNode *) mainData;
+            THROW_FMT(FormatError, "unknown Sequence: %d", info);
+
+        case RM7_BITMAP_ID:
+            switch (info)
+            {
+                case XLOG_BITMAP_INSERT_WORDS:
+                case XLOG_BITMAP_UPDATEWORD:
+                case XLOG_BITMAP_UPDATEWORDS:
+                case XLOG_BITMAP_INSERT_LOVITEM:
+                case XLOG_BITMAP_INSERT_BITMAP_LASTWORDS:
+                case XLOG_BITMAP_INSERT_META:
+                    return (const RelFileNode *) mainData;
+
+                default:
+                    break;
+            }
+            break;
+
+        case RM7_APPEND_ONLY_ID:
+            switch (info)
+            {
+                case XLOG_APPENDONLY_INSERT:
+                case XLOG_APPENDONLY_TRUNCATE:
+                    return (const RelFileNode *) mainData;
+
+                default:
+                    THROW_FMT(FormatError, "unknown Appendonly: %d", info);
+            }
+    }
+
+    return NULL;
+}
+
+static List *
+getRelFileNodes(XLogRecordGPDB7 *const record, PgPageSize pageSize)
+{
+    #define COPY_HEADER_FIELD(_dst, _size) \
+    do                                     \
+    {                                      \
+        if (remaining < _size)             \
+            return NULL;                   \
+        memcpy(_dst, ptr, _size);          \
+        ptr += _size;                      \
+        remaining -= (uint32_t) _size;     \
+    }                                      \
+    while(0)
```

Comment:
COPY_HEADER_FIELD is always used in the form: `COPY_HEADER_FIELD(&a, sizeof(a))`, so we can simplify the macro and exclude mistakes like `COPY_HEADER_FIELD(&a, sizeof(b))`

```suggestion
    #define COPY_HEADER_FIELD(_dst)       \
    do                                    \
    {                                     \
        if (remaining < sizeof(_dst))     \
            return NULL;                  \
        memcpy(&_dst, ptr, sizeof(_dst)); \
        ptr += _size;                     \
        remaining -= sizeof(_dst);        \
    }                                     \
    while(0)
```